sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceExceptWithFilter.scala

```diff
@@ -0,0 +1,111 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import scala.annotation.tailrec
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.rules.Rule
+
+
+/**
+ * If one or both of the datasets in the logical [[Except]] operator are purely transformed using
+ * [[Filter]], this rule will replace logical [[Except]] operator with a [[Filter]] operator by
+ * flipping the filter condition of the right child.
+ * {{{
+ *   SELECT a1, a2 FROM Tab1 WHERE a2 = 12 EXCEPT SELECT a1, a2 FROM Tab1 WHERE a1 = 5
+ *   ==>  SELECT DISTINCT a1, a2 FROM Tab1 WHERE a2 = 12 AND (a1 is null OR a1 <> 5)
+ * }}}
+ *
+ * Note:
+ * Before flipping the filter condition of the right node, we should:
+ * 1. Combine all it's [[Filter]].
+ * 2. Apply InferFiltersFromConstraints rule (to support NULL values of the condition).
+ */
+object ReplaceExceptWithFilter extends Rule[LogicalPlan] {
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case Except(left, right) if isEligible(left, right) =>
+      Distinct(Filter(Not(transformCondition(left, skipProject(right))), left))
+  }
+
+  private def transformCondition(left: LogicalPlan, right: LogicalPlan): Expression = {
+    val filterCondition = InferFiltersFromConstraints(combineFilters(right)
+    ).asInstanceOf[Filter].condition
+
+    val attributeNameMap: Map[String, Attribute] = left.output.map(x => (x.name, x)).toMap
+    val transformedCondition = filterCondition transform { case a : AttributeReference =>
+      attributeNameMap(a.name)
+    }
+
+    transformedCondition
+  }
+
+  private def isEligible(left: LogicalPlan, right: LogicalPlan): Boolean = (left, right) match {
+    case (_, right @ (Project(_, _: Filter) | Filter(_, _))) => verifyConditions(left, right)
+    case _ => false
+  }
+
+  private def verifyConditions(left: LogicalPlan, right: LogicalPlan): Boolean = {
+    val leftProjectList = projectList(left)
+    val rightProjectList = projectList(right)
+
+    left.output.size == left.output.distinct.size &&
+      verifyFilterCondition(skipProject(left)) && verifyFilterCondition(skipProject(right)) &&
+        Project(leftProjectList, nonFilterChild(skipProject(left))).sameResult(
+          Project(rightProjectList, nonFilterChild(skipProject(right))))
```

Comment:
How about this?

```Scala
    left.output.size == left.output.distinct.size &&
      left.expressions.forall(!SubqueryExpression.hasSubquery(_)) &&
      right.expressions.forall(!SubqueryExpression.hasSubquery(_)) &&
      Project(projectList(right), nonFilterChild(skipProject(left))).sameResult(
        Project(projectList(right), nonFilterChild(skipProject(right))))
```

Reply:
Then, `verifyFilterCondition ` and `collectAllExpressions ` can be removed.

Reply:
For the rule it is enough to test SubqueryExpressions for the top Filter node(s) conditions in the tree. I think `node.expressions` will return all the expressions of the entire tree, which is actually not needed. Correct me if i'm wrong..

Reply:
It would be nice if there is a function like `def takeWhile(p: A => Boolean)` in functional programming to substitute this logic in the verifyFilterCondition function:
```scala
    var i = 0
    val filterConditions = new collection.mutable.ArrayBuffer[Expression]
    while (plan.p(i).isInstanceOf[Filter]) {
      val condition = plan.p(i).asInstanceOf[Filter].condition
      filterConditions.insertAll(i, collectAllExpressions(condition))
      i += 1
    }
```

Reply:
As a second thought, i think it's possible to remove the `collectAllExpressions` function by modifying the `verifyFilterCondition` function like below. But assuming there could be another filter follows, we need to iterate down, so we still need `verifyFilterCondition `.
```scala
private def verifyFilterCondition(plan: LogicalPlan): Boolean = {
    var i = 0
    val filterConditions = new collection.mutable.ArrayBuffer[Expression]
    while (plan.p(i).isInstanceOf[Filter]) {
      filterConditions.insertAll(i, plan.p(i).expressions)
      i += 1
    }
    filterConditions.forall(!SubqueryExpression.hasSubquery(_))
  }
```

Reply:
And i didn't notice, the `left.output.size == left.output.distinct.size` in the verifyConditions function has to be changed something like 
`left.output.size == left.output.map(_.name).distinct.size`

Reply:
I missed it. `left.expressions` only returns the expressions of the left operator. It does not include the expressions of the children.

Reply:
Does `left.find(_.expressions.exists(SubqueryExpression.hasSubquery)).isEmpty` work?

Reply:
The point here is to avoid the extra complexity.

Reply:
@gatorsmile doing `left.find(_.expressions.exists(SubqueryExpression.hasSubquery)).isEmpty` will check for the `SubqueryExpression` for all the children of the left node. But it's enough to do the verification only for the successive Filter node(s).

Reply:
Let us check all the children of the left node. This will not be an issue for 99% cases and also simplify the codes.

Reply:
That makes sense.